File: src/tirx/ir/layout/layout.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "utils.h"

namespace tvm {
namespace tirx {

/**************** Layout ****************/
ffi::Map<ffi::String, PrimExpr> LayoutNode::Apply(const ffi::Array<PrimExpr>& coord,
                                                  const ffi::Array<PrimExpr>& shape) const {
  TVM_FFI_ICHECK_EQ(coord.size(), shape.size())
      << "ValueError: The size of coord and shape should be equal";
  return Apply(FlattenCoord(coord, shape));
}

TVM_FFI_STATIC_INIT_BLOCK() {
  namespace refl = tvm::ffi::reflection;
  auto def = refl::GlobalDef();
  def.def("tirx.LayoutCompatibleWithShape",
          [](Layout layout, Array<PrimExpr> shape) { return layout->CompatibleWithShape(shape); });
  def.def("tirx.LayoutVerifyWellFormed", [](Layout layout) { return layout->VerifyWellFormed(); });
  def.def("tirx.LayoutGetSize", [](Layout layout, ffi::Optional<ffi::String> axis_name) {
    return layout->GetSize(axis_name);
  });
  def.def("tirx.LayoutGetSpan", [](Layout layout, ffi::Optional<ffi::String> axis_name) {
    return layout->GetSpan(axis_name);
  });
  def.def("tirx.LayoutApplyWithShape",
          [](Layout layout, ffi::Array<PrimExpr> coord, ffi::Array<PrimExpr> shape) {
            return layout->Apply(coord, shape);
          });
  def.def("tirx.LayoutApply",
          [](Layout layout, ffi::Array<PrimExpr> coord) { return layout->Apply(coord); });
  def.def("tirx.LayoutApplyLinear",
          [](Layout layout, PrimExpr coord) { return layout->Apply(coord); });
  def.def("tirx.LayoutCanonicalize", [](Layout layout) { return layout->Canonicalize(); });
  def.def("tirx.LayoutTile", [](Layout layout, TileLayout outer, ffi::Array<PrimExpr> outer_shape,
                                ffi::Array<PrimExpr> inner_shape) {
    return layout->Tile(outer, outer_shape, inner_shape);
  });
  def.def("tirx.LayoutDirectSum",
          [](Layout layout, TileLayout left, ffi::Array<PrimExpr> left_shape,
             ffi::Array<PrimExpr> right_shape) {
            return layout->DirectSum(left, left_shape, right_shape);
          });
  def.def("tirx.LayoutIsTileInner",
          [](Layout layout, Layout tile_layout, ffi::Array<PrimExpr> tiled_shape,
             ffi::Array<PrimExpr> inner_shape) {
            return layout->IsTileInner(tile_layout, tiled_shape, inner_shape);
          });
  def.def("tirx.LayoutIsTileOuter",
          [](Layout layout, Layout tile_layout, ffi::Array<PrimExpr> tiled_shape,
             ffi::Array<PrimExpr> outer_shape) {
            return layout->IsTileOuter(tile_layout, tiled_shape, outer_shape);
          });
  def.def("tirx.LayoutIsDirectSumRight",
          [](Layout layout, Layout sum_layout, ffi::Array<PrimExpr> interleaved_shape,
             ffi::Array<PrimExpr> right_shape) {
            return layout->IsDirectSumRight(sum_layout, interleaved_shape, right_shape);
          });
  def.def("tirx.LayoutIsDirectSumLeft",
          [](Layout layout, Layout sum_layout, ffi::Array<PrimExpr> interleaved_shape,
             ffi::Array<PrimExpr> left_shape) {
            return layout->IsDirectSumLeft(sum_layout, interleaved_shape, left_shape);
          });
  def.def("tirx.LayoutSlice",
          [](Layout layout, ffi::Array<PrimExpr> shape, Region region) -> ffi::Optional<Layout> {
            return layout->Slice(shape, region);
          });
}

}  // namespace tirx
}  // namespace tvm

File: python/tvm/backend/trn/layout.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""Trainium-specific TIRx layout helpers."""

from __future__ import annotations

import functools
import operator
import re

import tvm
from tvm.tirx.expr import PrimExpr
from tvm.tirx.layout import Axis, Iter, Layout, S, TileLayout

_TRN_MEMORY_AXES = {"F", "P", "Bank"}
_PSUM_MAX_ELEM_PER_BANK = 512


def is_trainium_layout(layout: Layout | None) -> bool:
    """Return whether a layout uses only Trainium memory axes."""
    if not isinstance(layout, TileLayout):
        return False
    return not any(
        iter.axis.is_memory() and iter.axis.name not in _TRN_MEMORY_AXES for iter in layout.shard
    )


def trainium_layout(annotation: str, shape: tuple[PrimExpr], is_psum: bool = False) -> TileLayout:
    """Create a Trainium tile layout from a PF annotation string and logical shape."""
    analyzer = tvm.arith.Analyzer()
    assert re.fullmatch(r"[PF]*", annotation), (
        f"annotation {annotation} must be a string of 'P' and 'F'"
    )
    assert len(annotation) == len(shape), (
        f"annotation {annotation} and shape {shape} must have the same length"
    )
    num_p_dim = annotation.count("P")
    if num_p_dim == 1:
        p_idx = annotation.index("P")
        p_dim = shape[p_idx]
        assert analyzer.can_prove(p_dim <= 128 or p_dim % 128 == 0), (
            f"There is only 1 P in the annotation. Partition size {p_dim} must be less than "
            "or equal to 128 or a multiple of 128"
        )
        if analyzer.can_prove(p_dim > 128):
            annotation = "F" + annotation
            shape = (p_dim // 128, *shape[:p_idx], 128, *shape[p_idx + 1 :])
    elif num_p_dim > 1:
        p_dim_prod = functools.reduce(
            operator.mul, [s for s, c in zip(shape, annotation) if c == "P"]
        )
        assert analyzer.can_prove(p_dim_prod <= 128), (
            f"There are {num_p_dim} Ps in the annotation. Partition size {p_dim_prod} must be "
            "less than or equal to 128"
        )

    f_shape = [s for s, c in zip(shape, annotation) if c == "F"]
    p_shape = [s for s, c in zip(shape, annotation) if c == "P"]
    f_strides = Layout._get_default_strides(f_shape, 1)  # pylint: disable=protected-access
    p_strides = Layout._get_default_strides(p_shape, 1)  # pylint: disable=protected-access
    f_tile_layout = TileLayout(S[tuple(f_shape) : tuple(s @ Axis.F for s in f_strides)])
    p_tile_layout = TileLayout(S[tuple(p_shape) : tuple(s @ Axis.P for s in p_strides)])
    result = []
    f_index = p_index = 0

    for char in annotation:
        if char == "F":
            result.append(f_tile_layout.shard[f_index])
            f_index += 1
        else:
            result.append(p_tile_layout.shard[p_index])
            p_index += 1
    if num_p_dim == 1 and analyzer.can_prove(p_dim > 128):
        higher_p = result[0]
        result = result[1:]
        result = [*result[:p_idx], higher_p, *result[p_idx:]]

    res = TileLayout.from_iters(result, [], {})
    if is_psum:
        res = to_psum_layout(res)
    return res


def to_psum_layout(layout: TileLayout) -> TileLayout:
    """Convert a Trainium sbuf layout to its psum physical-bank layout."""
    analyzer = tvm.arith.Analyzer()
    shard = []
    for iter in layout.shard:
        if iter.axis.name == "F":
            if analyzer.can_prove(iter.stride % _PSUM_MAX_ELEM_PER_BANK == 0):
                stride = analyzer.simplify(iter.stride // _PSUM_MAX_ELEM_PER_BANK)
                shard.append(Iter(iter.extent, stride, Axis.get("Bank")))
            elif analyzer.can_prove(_PSUM_MAX_ELEM_PER_BANK % iter.stride == 0):
                c = analyzer.simplify(_PSUM_MAX_ELEM_PER_BANK // iter.stride)
                if analyzer.can_prove(iter.extent < c):
                    shard.append(iter)
                elif analyzer.can_prove(iter.extent % c == 0):
                    shard.append(Iter(analyzer.simplify(iter.extent // c), 1, Axis.get("Bank")))
                    shard.append(Iter(c, iter.stride, Axis.get("F")))
                else:
                    raise ValueError(f"layout {layout} can not be converted to psum layout")
            else:
                raise ValueError(f"layout {layout} can not be converted to psum layout")
        else:
            shard.append(iter)
    return TileLayout.from_iters(shard, [], {})


__all__ = ["is_trainium_layout", "to_psum_layout", "trainium_layout"]
